pkg/cmd/run/view/view.go

```diff
@@ -453,57 +424,38 @@ func promptForJob(cs *iostreams.ColorScheme, jobs []shared.Job) (*shared.Job, er
 // └── jobname2/
 //     ├── 1_stepname.txt
 //     └── 2_somestepname.txt
-func readRunLog(rlz io.ReadCloser) (runLog, error) {
-	rl := make(runLog)
+// It iterates through the list of jobs and trys to find the matching
+// log in the zip file. If the matching log is found it is attached
+// to the job.
+func readRunLog(rlz io.ReadCloser, jobs []shared.Job) error {
 	defer rlz.Close()
 	z, err := ioutil.ReadAll(rlz)
 	if err != nil {
-		return rl, err
+		return err
 	}
 
 	zipReader, err := zip.NewReader(bytes.NewReader(z), int64(len(z)))
 	if err != nil {
-		return rl, err
+		return err
 	}
 
-	for _, zipFile := range zipReader.File {
-		dir, file := filepath.Split(zipFile.Name)
-		ext := filepath.Ext(zipFile.Name)
-
-		// Skip all top level files and non-text files
-		if dir != "" && ext == ".txt" {
-			split := strings.Split(file, "_")
-			if len(split) != 2 {
-				return rl, errors.New("invalid step log filename")
-			}
-
-			jobName := strings.TrimSuffix(dir, "/")
-			stepName := strings.TrimSuffix(split[1], ".txt")
-			stepOrder, err := strconv.Atoi(split[0])
-			if err != nil {
-				return rl, errors.New("invalid step log filename")
-			}
-
-			stepLogs, err := readZipFile(zipFile)
-			if err != nil {
-				return rl, err
-			}
-
-			st := step{
-				order: stepOrder,
-				name:  stepName,
-				logs:  string(stepLogs),
-			}
-
-			if j, ok := rl[jobName]; !ok {
-				rl[jobName] = &job{name: jobName, steps: []step{st}}
-			} else {
-				j.steps = append(j.steps, st)
+	for i, job := range jobs {
+		for j, step := range job.Steps {
+			filename := fmt.Sprintf("%s/%d_%s.txt", job.Name, step.Number, step.Name)
+			for _, file := range zipReader.File {
+				if file.Name == filename {
+					log, err := readZipFile(file)
+					if err != nil {
+						return err
+					}
+					jobs[i].Steps[j].Log = string(log)
```

Comment:
Logs can be big! For the sake of memory, maybe it would be better to structure this code to write to a Writer instead of saving it to a field as a string.

Same goes for the zip file itself: right now it's read to memory, but it could be written to a temporary file instead. Plus, if we go the filesystem route, we could avoid re-downloading a zip file if we detect that it's already been downloaded in a previous invocation.

Reply:
Great suggestions. I added in some caching and changed the approach of attaching logs as suggested.